node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -106,7 +107,7 @@ retrievalWorkerImpl SendActions {..} =
                 (_, Just (nodeId, rHeader))  ->
                     pure (handleRecoveryWithHandler nodeId rHeader)
         thingToDoNext
-        mainLoop
+        runIfNotShutdown mainLoopImpl -- next iteration
```

Comment:
Do we need `runIfNotShutdown`? Can't we just `killThread` to stop this?

Reply:
Current shutdown machinery uses `runIfNotShutdown` approach. If it's going to be changed, it'll be the separate issue/PR. I think you can ask @gromakovsky about it.

Reply:
`runIfNotShutdown` is an extremely obsolete thing. It's dead for months, but this PR is to `master` branch (which is extremely old).